[ru] update 'Web/JavaScript/Reference/Operators/in' translation (#21442)
mdn/translated-content: files/ru/web/javascript/reference/operators/in/index.md

---
title: in
slug: Web/JavaScript/Reference/Operators/in

> ---

Reviewer: ```suggestion
l10n:
  sourceCommit: 13b2fe46243beed6dc29f5d3defa716abe79c1f9
---
```

---
title: in
slug: Web/JavaScript/Reference/Operators/in
l10n:
  sourceCommit: 13b2fe46243beed6dc29f5d3defa716abe79c1f9
---

{{jsSidebar("Operators")}}

Оператор `in` возвращает `true`, если свойство содержится в указанном объекте или в его цепочке прототипов.

> `in` не может быть использован для поиска значений в других коллекциях. Чтобы проверить, существует ли определенное значение в массиве, используйте {{jsxref("Array.prototype.includes()")}}. Для Set используйте {{jsxref("Set.prototype.has()")}}.

Reviewer: ```suggestion
Оператор `in` не может быть использован для поиска значений в других видах коллекций. Чтобы проверить, существует ли определённое значение в массиве, можно использовать {{jsxref("Array.prototype.includes()")}}. А у наборов есть метод {{jsxref("Set.prototype.has()")}}.
```

{{EmbedInteractiveExample("pages/js/expressions-inoperator.html")}}

## Синтаксис

```js-nolint
prop in object
#prop in object
```

### Параметры

- `prop`

>   - : Строка или symbol, представляющий название свойства (non-symbols будут [конвертированы в строки](/ru/docs/Web/JavaScript/Reference/Global_Objects/String#string_coercion)). Также может быть [идентификатором приватного свойства](/ru/docs/Web/JavaScript/Reference/Classes/Private_properties).

Reviewer: ```suggestion
  - : Строковое или символьное значение, представляющее название свойства (несимвольные значения будут [преобразованы в строки](/ru/docs/Web/JavaScript/Reference/Global_Objects/String#приведение_к_строке)). Также может быть [именем приватного свойства](/ru/docs/Web/JavaScript/Reference/Classes/Private_properties).
```

- `object`

>   - : Объект, в котором нужно проверить (или в его цепочке прототипов) содержание свойства с заданным именем (`prop`).

Reviewer: ```suggestion
  - : Объект, для которого будет производится проверка, содержит ли он (или его цепочка прототипов) свойство с указанным именем (`prop`).
```

Свойство может существовать в объекте, но иметь значение `undefined`. Поэтому `x in obj` не то же самое, что `obj.x !== undefined`. Чтобы `in` возвращал значение `false` после того, как свойство было добавлено, используйте оператор [`delete`](/ru/docs/Web/JavaScript/Reference/Operators/delete), вместо присваивания свойству `undefined`.

Вы также можете использовать оператор `in`, чтобы проверить, существует ли в объекте [приватное поле класса или метод](/ru/docs/Web/JavaScript/Reference/Classes/Private_properties). Оператор возвращает `true` если свойство определено, в противном случае `false`. This is known as a _branded check_, потому что `in` возвращает `true` тогда и только тогда, когда объект был создан с помощью конструктора класса, после чего вы можете безопасно получить доступ к другим приватным свойствам.

Проверка приватных свойств оператором `in` это особый синтаксис - левая сторона оператора `in` является идентификатором свойства, а не выражением, но без кавычек (иначе это является свойством с типом строка, а не приватным свойством).

> Поскольку при обращении к приватным свойствам объекта не связанного с текущим классом выбрасывает {{jsxref("TypeError")}} вместо возврата `undefined`, то оператор `in` позволят сократить запись:

Author: На странице об [Error](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Error) - используется глагол "выбрасывается" по отношению к ошибкам. Почему ты предлагаешь использовать "вызывает"?

> Поскольку при обращении к приватным свойствам объекта не связанного с текущим классом выбрасывает {{jsxref("TypeError")}} вместо возврата `undefined`, то оператор `in` позволят сократить запись:

Reviewer: Постараюсь объяснить свою точку зрения)
Разные формы слов "выбрасывать" действительно часто используются в MDN (и в других источниках тоже), но обычно в сочетании с "исключениями", то есть можно сказать есть такие устойчивые выражаения ("выброс исключения", "выбрасывать исключение"). Мне такие сочетания кажутся не лучшим вариантом перевода, потому что по смыслу исключение (или ошибка) скорее "возникает", "появляется". А слово "выбрасывать" имеет двойственное толкование: выбрасывать можно мусор в урну, а можно выбросить товар на рынок (или на всеобщее обозрение). Применительно к ошибкам мы хотим использовать второе значение, конечно, но почему бы не избегать этой дополнительной когнитивной нагрузки?
Второй момент: с точки зрения языка и логики исключение выбрасывает программа, а не "обращение к приватным свойствам". Чтобы не удлинять предложение и не упоминать программу (среду выполнения), можно использовать здесь слово "выбрасывается" (кстати вот и третий пример значения слова: выбрасываться можно как кит на берег).

Сейчас пришёл ещё более корректный и приятный для чтения вариант: "приводит к появлению", предлагаю его и применить)

```js
class C {
  #x;
  static isC(obj) {
    try {
      obj.#x;
      return true;
    } catch {
      return false;
    }
  }
}
```

Короткая запись:

```js
class C {
  #x;
  static isC(obj) {
    return #x in obj;
  }
}
```

Оператор `in` обычно позволяет избежать обработки ошибок только для работы с приватным свойством, которое может отсутствовать.

> Однако, оператор `in` по-прежнему требует предварительно объявлять приватные свойства заранее в окружающем классе, иначе будет выброшена ошибка {{jsxref("SyntaxError")}} ("Private field '#x' must be declared in an enclosing class") такая же, как и когда вы пытаетесь получить доступ к необъявленному приватному свойству.

Reviewer: ```suggestion
Однако оператор `in` по-прежнему требует предварительного объявления приватных свойств в окружающем классе, иначе возникает ошибка {{jsxref("SyntaxError")}} ("Private field '#x' must be declared in an enclosing class") такая же, как и при попытке получить доступ к необъявленному приватному свойству.
```

```js-nolint example-bad
class C {
  foo() {
    #x in this;
  }
}

new C().foo(); // SyntaxError: Private field '#x' must be declared in an enclosing class
```

## Примеры

> ### Стандартное использование

Reviewer: ```suggestion
### Обычное использование
```

В примере ниже показаны некоторые способы использования оператора `in`.

```js
// Массивы

> const trees = new Array("redwood", "bay", "cedar", "oak", "maple");

Reviewer: ```suggestion
const trees = ["redwood", "bay", "cedar", "oak", "maple"];
```

0 in trees; // true
3 in trees; // true
6 in trees; // false

> "bay" in trees; // false (вы должны указать индекс элемента в массиве, а не значение в этом индексе)

Reviewer: ```suggestion
"bay" in trees; // false (необходимо указать индекс элемента в массиве, а не значение)
```

"length" in trees; // true (length является свойством Array)
Symbol.iterator in trees; // true

// Уже существующие объекты
"PI" in Math; // true

// Пользовательские объекты
const mycar = { make: "Honda", model: "Accord", year: 1998 };
"make" in mycar; // true
"model" in mycar; // true
```

> Вы должны указать объект справа от оператора `in`. Например, вы можете указать строку, созданную через конструктор объекта `String`, но вы не можете указать строковый литерал.

Reviewer: ```suggestion
Необходимо указать объект справа от оператора `in`. Например, можно указать строку, созданную с помощью конструктора `String`, но нельзя использовать строковый литерал.
```

```js
const color1 = new String("green");

> "length" in color1; // returns true

Reviewer: ```suggestion
"length" in color1; // true
```

const color2 = "coral";

> // сгенерирует ошибку (color2 is not a String object)

Reviewer: ```suggestion
// сгенерирует ошибку (color2 не является объектом String)
```

> Если вы удалили свойство при помощи оператора [`delete`](/ru/docs/Web/JavaScript/Reference/Operators/delete), то оператор `in` возвратит `false` для этого свойства.

Reviewer: ```suggestion
Если удалить свойство при помощи оператора [`delete`](/ru/docs/Web/JavaScript/Reference/Operators/delete), то оператор `in` возвратит `false` для этого свойства.
```

```js
const mycar = { make: "Honda", model: "Accord", year: 1998 };
delete mycar.make;
"make" in mycar; // false

> const trees = new Array("redwood", "bay", "cedar", "oak", "maple");

Reviewer: ```suggestion
const trees = ["redwood", "bay", "cedar", "oak", "maple"];
```
